design/repository-maintenance.md

```diff
@@ -0,0 +1,236 @@
+		VolumeMounts(*builder.ForVolumeMount("cloud-credentials", "/credentials").Result()).
+		ServiceAccount("velero").Resources(&resources).
+		Volumes(*builder.ForVolume("cloud-credentials").SecretSource("cloud-credentials").Result()).Result()
+	...
+	// Maintenance job also inherits some other filed values from the Velero server pod in keeping with the setting of the Velero server pod
+	log.Debugf("Creating maintenance job: %v", job)
+	defer cancel()
+	if err := m.client.Create(pruneTimeOut, job); err != nil {
+		return errors.Wrap(err, "error to create maintenance job")
+	}
+
+	if err := builder.WaitForJobComplete(pruneTimeOut, m.client, job); err != nil {
+		return errors.Wrap(err, "error to wait for maintenance job complete")
+	}
+	...
+	return nil
+}
+```
+
+Now, the backup repository controller will call the repository manager to create one maintenance job and wait for the job to complete. The Kopia or Restic maintenance multiple chains are called by the job.
+
+### 3. Update the Result of the Maintenance Job into BackupRepository CR
+
+The backup repository controller will update the result of the maintenance job into the backup repository CR.
+
+For how to get the result of the maintenance job we could refer to [here](https://kubernetes.io/docs/tasks/debug/debug-application/determine-reason-pod-failure/#writing-and-reading-a-termination-message).
+
+After the maintenance job is finished, we could get the result of maintenance by getting the terminated message from the related pod:
+
+```golang
+func GetContainerTerminatedMessage(pod *v1.Pod) string {
+	...
+	for _, containerStatus := range pod.Status.ContainerStatuses {
+		if containerStatus.LastTerminationState.Terminated != nil {
+			return containerStatus.LastTerminationState.Terminated.Message
+		}
+	}
+	...
+	return ""
+}
+```
+Then we could update the status of backupRepository CR with the message.
+
+### 4. Add Setting for Maintenance Job Timeout
+Add one configuration for setting maintenance job timeout as below:
+```shell
+    velero server --maintenance-job-timeout $maintenance-job-timeout
+```
+Our default value is 0, which means we don't limit the timeout duration.
+
+### 5. Add Setting for Resource Usage of Maintenance
+Add one configuration for setting the resource limit of maintenance jobs as below:
+```shell
+    velero server --maintenance-job-cpu-request $cpu-request --maintenance-job-mem-request $mem-request --maintenance-job-cpu-limit $cpu-limit --maintenance-job-mem-limit $mem-limit
+```
+Our default value is 0, which means we don't limit the resources.
+
+### 6. Clear up Maintenance Jobs
+Add one configuration for the number of keeping latest failure maintenance jobs as below:
+```shell
+    velero server --keep-failed-maintenance-jobs $num
+```
+
+We would clean all completed jobs and keep the latest N failed jobs at the end of each round of reconciliation.
+
+```golang
+func clearFinishedJob(ctx context.Context, c client.Client, retainFailedJobs int) error {
```

Comment:
When will the clean-up action be triggered?

Reply:
it will triggered at the end of Reconcile, will it be so frequently?

Reply:
If it is triggered at the end of reconciling, does that mean any change of the repository also triggers the clean-up action? Seems this is unnecessary. 
How about triggering the clean-up action to clean up the corresponding repository's jobs when one job completes?

Reply:
ok, I think it's more reasonable